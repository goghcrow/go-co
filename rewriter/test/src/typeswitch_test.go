package src

import (
	"testing"

	. "github.com/goghcrow/go-co"
)

func TestTypeSwitchTrival(t *testing.T) {
	f := func() {}
	g := func() (_ Iter[int]) {
		Yield(0)
		var a any
		switch a.(type) {
		case int:
			Yield(1)
		case string:
			Yield(2)
		default:
			f()
			Yield(3)
			f()
		}
		Yield(4)
		f()
		return
	}
	xs := iter2slice(g())
	assertEqual(t, xs, []int{0, 3, 4})
}

func TestTypeSwitchTrivalInitAndNotDefaultBranch(t *testing.T) {
	g := func(a any) (_ Iter[int]) {
		Yield(0)
		switch x := a; x.(type) {
		case int:
			Yield(1)
		case string:
			Yield(2)
		default:
			Yield(3)
		}
		Yield(4)
		return
	}

	{
		xs := iter2slice(g(1))
		assertEqual(t, xs, []int{0, 1, 4})
	}
	{
		xs := iter2slice(g(""))
		assertEqual(t, xs, []int{0, 2, 4})
	}
	{
		xs := iter2slice(g(3.13))
		assertEqual(t, xs, []int{0, 3, 4})
	}
}

func TestTypeSwitchScope(t *testing.T) {
	g := func(a any) (_ Iter[int]) {
		Yield(0)
		x := 42
		switch x := a; x.(type) {
		case int:
			Yield(1)
		case string:
			Yield(2)
		default:
			Yield(3)
		}
		Yield(x)
		return
	}

	{
		xs := iter2slice(g(1))
		assertEqual(t, xs, []int{0, 1, 42})
	}
	{
		xs := iter2slice(g(""))
		assertEqual(t, xs, []int{0, 2, 42})
	}
	{
		xs := iter2slice(g(3.13))
		assertEqual(t, xs, []int{0, 3, 42})
	}
}

func TestTypeSwitchYieldInitAndNotDefaultBranch(t *testing.T) {
	g := func() (_ Iter[int]) {
		var a any
		switch Yield(0); a.(type) {
		case int:
			Yield(1)
		case string:
			Yield(2)
		}
		Yield(3)
		return
	}
	xs := iter2slice(g())
	assertEqual(t, xs, []int{0, 3})
}

func TestTypeSwitchYieldInitAndNotDefaultBranch1(t *testing.T) {
	g := func() (_ Iter[int]) {
		var a any = "hello"
		switch Yield(0); a.(type) {
		case int:
			Yield(1)
		case string:
			Yield(2)
		}
		Yield(3)
		return
	}
	xs := iter2slice(g())
	assertEqual(t, xs, []int{0, 2, 3})
}

func TestTypeSwitchYieldInitAndNotDefaultBranch2(t *testing.T) {
	g := func() (_ Iter[int]) {
		var a any = "hello"
		switch Yield(0); a.(type) {
		case int:
			Yield(1)
		case string:
			Yield(2)
			return
		}
		Yield(3)
		return
	}
	xs := iter2slice(g())
	assertEqual(t, xs, []int{0, 2})
}

func TestTypeSwitchYieldInitAndAssign(t *testing.T) {
	g := func() (_ Iter[int]) {
		var a any = 42
		switch Yield(0); b := a.(type) {
		case int:
			Yield(b)
			return
		case string:
			Yield(2)
		}
		Yield(3)
		return
	}
	xs := iter2slice(g())
	assertEqual(t, xs, []int{0, 42})
}
