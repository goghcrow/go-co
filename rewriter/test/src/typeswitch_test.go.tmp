//go:build !co

// Code generated by github.com/goghcrow/go-co DO NOT EDIT.
package src

import (
	"testing"

	. "github.com/goghcrow/go-co"
	ʂɘʠ "github.com/goghcrow/go-co/seq"
)

func TestTypeSwitchTrival(t *testing.T) {
	f := func() {}
	g := func() (_ ʂɘʠ.Iterator[int]) {
		return ʂɘʠ.Start[int](ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {
			return ʂɘʠ.Bind[int](0, func() ʂɘʠ.Seq[int] {

				var a any
				return ʂɘʠ.Combine[int](ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {
					switch a.(type) {
					case int:
						return ʂɘʠ.Bind[int](1, func() ʂɘʠ.Seq[int] {
							return ʂɘʠ.Normal[int]()
						})
					case string:
						return ʂɘʠ.Bind[int](2, func() ʂɘʠ.Seq[int] {
							return ʂɘʠ.Normal[int]()
						})
					default:

						f()
						return ʂɘʠ.Bind[int](3, func() ʂɘʠ.Seq[int] {

							f()
							return ʂɘʠ.Normal[int]()
						})
					}
				}), ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {
					return ʂɘʠ.Bind[int](4, func() ʂɘʠ.Seq[int] {

						f()
						return ʂɘʠ.Return[int]()
					})
				}))
			})
		}))

	}
	xs := iter2slice(g())
	assertEqual(t, xs, []int{0, 3, 4})
}

func TestTypeSwitchTrivalInitAndNotDefaultBranch(t *testing.T) {
	g := func(a any) (_ ʂɘʠ.Iterator[int]) {
		return ʂɘʠ.Start[int](ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {
			return ʂɘʠ.Bind[int](0, func() ʂɘʠ.Seq[int] {
				return ʂɘʠ.Combine[int](ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {
					return ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {

						x := a
						switch x.(type) {
						case int:
							return ʂɘʠ.Bind[int](1, func() ʂɘʠ.Seq[int] {
								return ʂɘʠ.Normal[int]()
							})
						case string:
							return ʂɘʠ.Bind[int](2, func() ʂɘʠ.Seq[int] {
								return ʂɘʠ.Normal[int]()
							})
						default:
							return ʂɘʠ.Bind[int](3, func() ʂɘʠ.Seq[int] {
								return ʂɘʠ.Normal[int]()
							})
						}
					})
				}), ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {
					return ʂɘʠ.Bind[int](4, func() ʂɘʠ.Seq[int] {
						return ʂɘʠ.Return[int]()
					})
				}))
			})
		}))

	}

	{
		xs := iter2slice(g(1))
		assertEqual(t, xs, []int{0, 1, 4})
	}
	{
		xs := iter2slice(g(""))
		assertEqual(t, xs, []int{0, 2, 4})
	}
	{
		xs := iter2slice(g(3.13))
		assertEqual(t, xs, []int{0, 3, 4})
	}
}

func TestTypeSwitchScope(t *testing.T) {
	g := func(a any) (_ ʂɘʠ.Iterator[int]) {
		return ʂɘʠ.Start[int](ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {
			return ʂɘʠ.Bind[int](0, func() ʂɘʠ.Seq[int] {

				x := 42
				return ʂɘʠ.Combine[int](ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {
					return ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {

						x := a
						switch x.(type) {
						case int:
							return ʂɘʠ.Bind[int](1, func() ʂɘʠ.Seq[int] {
								return ʂɘʠ.Normal[int]()
							})
						case string:
							return ʂɘʠ.Bind[int](2, func() ʂɘʠ.Seq[int] {
								return ʂɘʠ.Normal[int]()
							})
						default:
							return ʂɘʠ.Bind[int](3, func() ʂɘʠ.Seq[int] {
								return ʂɘʠ.Normal[int]()
							})
						}
					})
				}), ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {
					return ʂɘʠ.Bind[int](x, func() ʂɘʠ.Seq[int] {
						return ʂɘʠ.Return[int]()
					})
				}))
			})
		}))

	}

	{
		xs := iter2slice(g(1))
		assertEqual(t, xs, []int{0, 1, 42})
	}
	{
		xs := iter2slice(g(""))
		assertEqual(t, xs, []int{0, 2, 42})
	}
	{
		xs := iter2slice(g(3.13))
		assertEqual(t, xs, []int{0, 3, 42})
	}
}

func TestTypeSwitchYieldInitAndNotDefaultBranch(t *testing.T) {
	g := func() (_ ʂɘʠ.Iterator[int]) {
		return ʂɘʠ.Start[int](ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {
			var a any
			return ʂɘʠ.Bind[int](0, func() ʂɘʠ.Seq[int] {
				return ʂɘʠ.Combine[int](ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {
					switch a.(type) {
					case int:
						return ʂɘʠ.Bind[int](1, func() ʂɘʠ.Seq[int] {
							return ʂɘʠ.Normal[int]()
						})
					case string:
						return ʂɘʠ.Bind[int](2, func() ʂɘʠ.Seq[int] {
							return ʂɘʠ.Normal[int]()
						})
					}
					return ʂɘʠ.Normal[int]()
				}), ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {
					return ʂɘʠ.Bind[int](3, func() ʂɘʠ.Seq[int] {
						return ʂɘʠ.Return[int]()
					})
				}))
			})
		}))

	}
	xs := iter2slice(g())
	assertEqual(t, xs, []int{0, 3})
}

func TestTypeSwitchYieldInitAndNotDefaultBranch1(t *testing.T) {
	g := func() (_ ʂɘʠ.Iterator[int]) {
		return ʂɘʠ.Start[int](ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {
			var a any = "hello"
			return ʂɘʠ.Bind[int](0, func() ʂɘʠ.Seq[int] {
				return ʂɘʠ.Combine[int](ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {
					switch a.(type) {
					case int:
						return ʂɘʠ.Bind[int](1, func() ʂɘʠ.Seq[int] {
							return ʂɘʠ.Normal[int]()
						})
					case string:
						return ʂɘʠ.Bind[int](2, func() ʂɘʠ.Seq[int] {
							return ʂɘʠ.Normal[int]()
						})
					}
					return ʂɘʠ.Normal[int]()
				}), ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {
					return ʂɘʠ.Bind[int](3, func() ʂɘʠ.Seq[int] {
						return ʂɘʠ.Return[int]()
					})
				}))
			})
		}))

	}
	xs := iter2slice(g())
	assertEqual(t, xs, []int{0, 2, 3})
}

func TestTypeSwitchYieldInitAndNotDefaultBranch2(t *testing.T) {
	g := func() (_ ʂɘʠ.Iterator[int]) {
		return ʂɘʠ.Start[int](ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {
			var a any = "hello"
			return ʂɘʠ.Bind[int](0, func() ʂɘʠ.Seq[int] {
				return ʂɘʠ.Combine[int](ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {
					switch a.(type) {
					case int:
						return ʂɘʠ.Bind[int](1, func() ʂɘʠ.Seq[int] {
							return ʂɘʠ.Normal[int]()
						})
					case string:
						return ʂɘʠ.Bind[int](2, func() ʂɘʠ.Seq[int] {
							return ʂɘʠ.Return[int]()
						})
					}
					return ʂɘʠ.Normal[int]()
				}), ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {
					return ʂɘʠ.Bind[int](3, func() ʂɘʠ.Seq[int] {
						return ʂɘʠ.Return[int]()
					})
				}))
			})
		}))

	}
	xs := iter2slice(g())
	assertEqual(t, xs, []int{0, 2})
}

func TestTypeSwitchYieldInitAndAssign(t *testing.T) {
	g := func() (_ ʂɘʠ.Iterator[int]) {
		return ʂɘʠ.Start[int](ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {
			var a any = 42
			return ʂɘʠ.Bind[int](0, func() ʂɘʠ.Seq[int] {
				return ʂɘʠ.Combine[int](ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {
					switch b := a.(type) {
					case int:
						return ʂɘʠ.Bind[int](b, func() ʂɘʠ.Seq[int] {
							return ʂɘʠ.Return[int]()
						})
					case string:
						return ʂɘʠ.Bind[int](2, func() ʂɘʠ.Seq[int] {
							return ʂɘʠ.Normal[int]()
						})
					}
					return ʂɘʠ.Normal[int]()
				}), ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {
					return ʂɘʠ.Bind[int](3, func() ʂɘʠ.Seq[int] {
						return ʂɘʠ.Return[int]()
					})
				}))
			})
		}))

	}
	xs := iter2slice(g())
	assertEqual(t, xs, []int{0, 42})
}
