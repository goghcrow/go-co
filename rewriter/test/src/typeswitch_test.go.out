//go:build !co

// Code generated by github.com/goghcrow/go-co DO NOT EDIT.
package src

import (
	ʂɘʠ "github.com/goghcrow/go-co/seq"
	"testing"
)

func TestTypeSwitchTrival(t *testing.T) {
	f := func() {}
	g := func() (_ ʂɘʠ.Iterator[int]) {
		return ʂɘʠ.Start[int](
			ʂɘʠ.Bind[int](0, func() ʂɘʠ.Seq[int] {

				var a any
				return ʂɘʠ.Combine[int](ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {
					switch a.(type) {
					case int:
						return ʂɘʠ.Bind[int](1,
							ʂɘʠ.Normal[int],
						)
					case string:
						return ʂɘʠ.Bind[int](2,
							ʂɘʠ.Normal[int],
						)
					default:

						f()
						return ʂɘʠ.Bind[int](3, func() ʂɘʠ.Seq[int] {

							f()
							return ʂɘʠ.Normal[int]()
						})
					}
				}),
					ʂɘʠ.Bind[int](4, func() ʂɘʠ.Seq[int] {

						f()
						return ʂɘʠ.Return[int]()
					}),
				)
			}),
		)

	}
	xs := iter2slice(g())
	assertEqual(t, xs, []int{0, 3, 4})
}

func TestTypeSwitchTrivalInitAndNotDefaultBranch(t *testing.T) {
	g := func(a any) (_ ʂɘʠ.Iterator[int]) {
		return ʂɘʠ.Start[int](
			ʂɘʠ.Bind[int](0, func() ʂɘʠ.Seq[int] {
				return ʂɘʠ.Combine[int](
					ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {

						x := a
						switch x.(type) {
						case int:
							return ʂɘʠ.Bind[int](1,
								ʂɘʠ.Normal[int],
							)
						case string:
							return ʂɘʠ.Bind[int](2,
								ʂɘʠ.Normal[int],
							)
						default:
							return ʂɘʠ.Bind[int](3,
								ʂɘʠ.Normal[int],
							)
						}
					}),

					ʂɘʠ.Bind[int](4,
						ʂɘʠ.Return[int],
					),
				)
			}),
		)

	}

	{
		xs := iter2slice(g(1))
		assertEqual(t, xs, []int{0, 1, 4})
	}
	{
		xs := iter2slice(g(""))
		assertEqual(t, xs, []int{0, 2, 4})
	}
	{
		xs := iter2slice(g(3.13))
		assertEqual(t, xs, []int{0, 3, 4})
	}
}

func TestTypeSwitchScope(t *testing.T) {
	g := func(a any) (_ ʂɘʠ.Iterator[int]) {
		return ʂɘʠ.Start[int](
			ʂɘʠ.Bind[int](0, func() ʂɘʠ.Seq[int] {

				x := 42
				return ʂɘʠ.Combine[int](
					ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {

						x := a
						switch x.(type) {
						case int:
							return ʂɘʠ.Bind[int](1,
								ʂɘʠ.Normal[int],
							)
						case string:
							return ʂɘʠ.Bind[int](2,
								ʂɘʠ.Normal[int],
							)
						default:
							return ʂɘʠ.Bind[int](3,
								ʂɘʠ.Normal[int],
							)
						}
					}),
					ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {
						return ʂɘʠ.Bind[int](x,
							ʂɘʠ.Return[int],
						)
					}))
			}),
		)

	}

	{
		xs := iter2slice(g(1))
		assertEqual(t, xs, []int{0, 1, 42})
	}
	{
		xs := iter2slice(g(""))
		assertEqual(t, xs, []int{0, 2, 42})
	}
	{
		xs := iter2slice(g(3.13))
		assertEqual(t, xs, []int{0, 3, 42})
	}
}

func TestTypeSwitchYieldInitAndNotDefaultBranch(t *testing.T) {
	g := func() (_ ʂɘʠ.Iterator[int]) {
		return ʂɘʠ.Start[int](ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {
			var a any
			return ʂɘʠ.Bind[int](0, func() ʂɘʠ.Seq[int] {
				return ʂɘʠ.Combine[int](ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {
					switch a.(type) {
					case int:
						return ʂɘʠ.Bind[int](1,
							ʂɘʠ.Normal[int],
						)
					case string:
						return ʂɘʠ.Bind[int](2,
							ʂɘʠ.Normal[int],
						)
					}
					return ʂɘʠ.Normal[int]()
				}),
					ʂɘʠ.Bind[int](3,
						ʂɘʠ.Return[int],
					),
				)
			})
		}))

	}
	xs := iter2slice(g())
	assertEqual(t, xs, []int{0, 3})
}

func TestTypeSwitchYieldInitAndNotDefaultBranch1(t *testing.T) {
	g := func() (_ ʂɘʠ.Iterator[int]) {
		return ʂɘʠ.Start[int](ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {
			var a any = "hello"
			return ʂɘʠ.Bind[int](0, func() ʂɘʠ.Seq[int] {
				return ʂɘʠ.Combine[int](ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {
					switch a.(type) {
					case int:
						return ʂɘʠ.Bind[int](1,
							ʂɘʠ.Normal[int],
						)
					case string:
						return ʂɘʠ.Bind[int](2,
							ʂɘʠ.Normal[int],
						)
					}
					return ʂɘʠ.Normal[int]()
				}),
					ʂɘʠ.Bind[int](3,
						ʂɘʠ.Return[int],
					),
				)
			})
		}))

	}
	xs := iter2slice(g())
	assertEqual(t, xs, []int{0, 2, 3})
}

func TestTypeSwitchYieldInitAndNotDefaultBranch2(t *testing.T) {
	g := func() (_ ʂɘʠ.Iterator[int]) {
		return ʂɘʠ.Start[int](ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {
			var a any = "hello"
			return ʂɘʠ.Bind[int](0, func() ʂɘʠ.Seq[int] {
				return ʂɘʠ.Combine[int](ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {
					switch a.(type) {
					case int:
						return ʂɘʠ.Bind[int](1,
							ʂɘʠ.Normal[int],
						)
					case string:
						return ʂɘʠ.Bind[int](2,
							ʂɘʠ.Return[int],
						)
					}
					return ʂɘʠ.Normal[int]()
				}),
					ʂɘʠ.Bind[int](3,
						ʂɘʠ.Return[int],
					),
				)
			})
		}))

	}
	xs := iter2slice(g())
	assertEqual(t, xs, []int{0, 2})
}

func TestTypeSwitchYieldInitAndAssign(t *testing.T) {
	g := func() (_ ʂɘʠ.Iterator[int]) {
		return ʂɘʠ.Start[int](ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {
			var a any = 42
			return ʂɘʠ.Bind[int](0, func() ʂɘʠ.Seq[int] {
				return ʂɘʠ.Combine[int](ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {
					switch b := a.(type) {
					case int:
						return ʂɘʠ.Bind[int](b,
							ʂɘʠ.Return[int],
						)
					case string:
						return ʂɘʠ.Bind[int](2,
							ʂɘʠ.Normal[int],
						)
					}
					return ʂɘʠ.Normal[int]()
				}),
					ʂɘʠ.Bind[int](3,
						ʂɘʠ.Return[int],
					),
				)
			})
		}))

	}
	xs := iter2slice(g())
	assertEqual(t, xs, []int{0, 42})
}
